### PR TITLE
[Modules] Adding new module for Microsoft.Insights/dataCollectionEndpoints

### LINKED ISSUE #2541: [Feature Request]: Add module for Microsoft.Insights/dataCollectionEndpoints
### Description

Currently there is no module for data collection endpoints (Microsoft.Insights/dataCollectionEndpoints), which provide a connection for certain data sources of Azure Monitor. The template reference can be found [here](https://learn.microsoft.com/en-us/azure/templates/microsoft.insights/datacollectionendpoints?pivots=deployment-language-bicep).

### PR DESCRIPTION
**Pull request by krbar**
# Description

This PR resolves #2541 and adds a new module for `Microsoft.Insights/dataCollectionEndpoints`.

| Pipeline |
| - |
| [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=132&branchName=users%2Fkrbar%2F2541-dataCollectionEndpointModule) |

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: modules/Microsoft.Insights/dataCollectionEndpoints/.test/common/dependencies.bicep [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+resource managedIdentity 'Microsoft.ManagedIdentity/userAssignedIdentities@2018-11-30' = {
+  name: managedIdentityName
+  location: location
+}
+
+@description('The principal ID of the created managed identity.')
+output managedIdentityPrincipalId string = managedIdentity.properties.principalId
+
+@description('The resource ID of the created Managed Identity.')
+output managedIdentityResourceId string = managedIdentity.id
```
Comment (AlexanderSehr) [inline modules/Microsoft.Insights/dataCollectionEndpoints/.test/common/dependencies.bicep:16]:
```suggestion
@description('The principal ID of the created managed identity.')
output managedIdentityPrincipalId string = managedIdentity.properties.principalId
```

As it appears that it is not used we can removed it

File: modules/Microsoft.Insights/dataCollectionEndpoints/.test/min/deploy.test.bicep [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+// ============== //
+// Test Execution //
+// ============== //
+
+module testDeployment '../../deploy.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name)}-test-${serviceShort}'
+  params: {
+    name: '<<namePrefix>>${serviceShort}001'
+  }
```
Comment (AlexanderSehr) [inline modules/Microsoft.Insights/dataCollectionEndpoints/.test/min/deploy.test.bicep:39]:
```suggestion
  params: {
    enableDefaultTelemetry: enableDefaultTelemetry
    name: '<<namePrefix>>${serviceShort}001'
  }
```

File: modules/Microsoft.Insights/dataCollectionEndpoints/readme.md [hunk 3]
```diff
@@ -0,0 +1,289 @@
+# Insights DataCollectionEndpoints `[Microsoft.Insights/dataCollectionEndpoints]`
+
+This module deploys Insights DataCollectionEndpoints.
+// TODO: Replace Resource and fill in description
```
Comment (AlexanderSehr) [inline modules/Microsoft.Insights/dataCollectionEndpoints/readme.md:4]:
```suggestion
This module deploys DataCollection Endpoints.
```

File: modules/Microsoft.Insights/dataCollectionEndpoints/readme.md [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+| `location` | string | `[resourceGroup().location]` |  | Location for all Resources. |
+| `lock` | string | `''` | `['', CanNotDelete, ReadOnly]` | Specify the type of lock. |
+| `publicNetworkAccess` | string | `'Disabled'` | `[Disabled, Enabled]` | The configuration to set whether network access from public internet to the endpoints are allowed. |
+| `roleAssignments` | array | `[]` |  | Array of role assignment objects that contain the 'roleDefinitionIdOrName' and 'principalId' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: '/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11'. |
+| `tags` | object | `{object}` |  | Resource tags. |
+
+
+### Parameter Usage: `<ParameterPlaceholder>`
+
+// TODO: Fill in Parameter usage
```
Comment (AlexanderSehr) [inline modules/Microsoft.Insights/dataCollectionEndpoints/readme.md:45]: Leftover

File: modules/Microsoft.Insights/dataCollectionEndpoints/deploy.bicep [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+resource dataCollectionEndpoint_lock 'Microsoft.Authorization/locks@2017-04-01' = if (!empty(lock)) {
+  name: '${dataCollectionEndpoint.name}-${lock}-lock'
+  properties: {
+    level: any(lock)
+    notes: lock == 'CanNotDelete' ? 'Cannot delete resource or child resources.' : 'Cannot modify the resource or child resources.'
+  }
+  scope: dataCollectionEndpoint
+}
+
+module dataCollectionRule_roleAssignments '.bicep/nested_roleAssignments.bicep' = [for (roleAssignment, index) in roleAssignments: {
```
Comment (AlexanderSehr) [inline modules/Microsoft.Insights/dataCollectionEndpoints/deploy.bicep:79]:
```suggestion
module dataCollectionEndpoint_roleAssignments '.bicep/nested_roleAssignments.bicep' = [for (roleAssignment, index) in roleAssignments: {
```

File: modules/Microsoft.Insights/dataCollectionEndpoints/readme.md (hunk 4 above)
Comment (krbar) [reply modules/Microsoft.Insights/dataCollectionEndpoints/readme.md]: fixed

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**